packages/dd-trace/src/exporters/agent/writer.js

```diff
@@ -1,129 +1,74 @@
-    this._lookup = lookup
+    this._appends = []
+    this._needsFlush = false
 
-    this._reset()
-  }
-
-  get length () {
-    return this._count
-  }
-
-  append (spans) {
-    log.debug(() => `Encoding trace: ${JSON.stringify(spans)}`)
-
-    this._encode(spans)
-  }
-
-  flush () {
-    if (this._count > 0) {
-      const data = platform.msgpack.prefix(this._buffer.slice(0, this._offset), this._count)
+    const fakeWriter = Object.create(BaseWriter.prototype)
+    fakeWriter._url = url
+    fakeWriter._prioritySampler = prioritySampler
+    fakeWriter._lookup = lookup
 
-      this._request(data, this._count)
-
-      this._reset()
-    }
-  }
-
-  _request (data, count) {
     const options = {
-      path: '/v0.4/traces',
+      path: '/v0.5/traces',
       method: 'PUT',
       headers: {
         'Content-Type': 'application/msgpack',
         'Datadog-Meta-Tracer-Version': tracerVersion,
-        'X-Datadog-Trace-Count': String(count)
+        'X-Datadog-Trace-Count': '0'
       },
-      lookup: this._lookup
+      lookup: fakeWriter._lookup
     }
 
-    this._setHeader(options.headers, 'Datadog-Meta-Lang', platform.name())
-    this._setHeader(options.headers, 'Datadog-Meta-Lang-Version', platform.version())
-    this._setHeader(options.headers, 'Datadog-Meta-Lang-Interpreter', platform.engine())
-
-    if (this._url.protocol === 'unix:') {
-      options.socketPath = this._url.pathname
+    fakeWriter._setHeader(options.headers, 'Datadog-Meta-Lang', platform.name())
+    fakeWriter._setHeader(options.headers, 'Datadog-Meta-Lang-Version', platform.version())
+    fakeWriter._setHeader(options.headers, 'Datadog-Meta-Lang-Interpreter', platform.engine())
+    if (fakeWriter._url.protocol === 'unix:') {
+      options.socketPath = fakeWriter._url.pathname
     } else {
-      options.protocol = this._url.protocol
-      options.hostname = this._url.hostname
-      options.port = this._url.port
+      options.protocol = fakeWriter._url.protocol
+      options.hostname = fakeWriter._url.hostname
+      options.port = fakeWriter._url.port
     }
 
-    log.debug(() => `Request to the agent: ${JSON.stringify(options)}`)
-
-    platform.metrics().increment(`${METRIC_PREFIX}.requests`, true)
-
-    platform.request(Object.assign({ data }, options), (err, res, status) => {
-      if (status) {
-        platform.metrics().increment(`${METRIC_PREFIX}.responses`, true)
-        platform.metrics().increment(`${METRIC_PREFIX}.responses.by.status`, `status:${status}`, true)
-      } else if (err) {
-        platform.metrics().increment(`${METRIC_PREFIX}.errors`, true)
-        platform.metrics().increment(`${METRIC_PREFIX}.errors.by.name`, `name:${err.name}`, true)
-
-        if (err.code) {
-          platform.metrics().increment(`${METRIC_PREFIX}.errors.by.code`, `code:${err.code}`, true)
-        }
+    const payload = { data: [msgpack.encode([[], []])] }
+    platform.request(Object.assign(payload, options), (err) => {
+      if (err) {
```

Comment:
What if the error is not because of version mismatch?

Reply:
My thinking was: All agents support 0.4, some agents support 0.5. Therefore, if there's any issue at all with 0.5, simply fall back to 0.4. Otherwise we'd need to do all the connection error handling logic in this meta-class, whose sole purpose is switch between the two writers. I figured this was a simpler approach.

Reply:
> What if the tracer was started before the agent for example?

Then it would immediately fall back to 0.4 and you'd have an unnecessarily degraded protocol. That's convinces me then. Will change.